### ISSUE TITLE
feat: multi-metric best checkpoint tracking — ensemble selection by validation metric

### ISSUE THREAD

**Issue by petteriTeikari** [petteriTeikari/minivess-mlops#893]
## Problem

During training, we save the best checkpoint for EACH tracked validation metric (best_val_loss,
best_val_dice, best_val_cldice, best_val_masd, best_val_compound_masd_cldice, last). This means
that for a 100-epoch training run, different metrics peak at different epochs:

- best_val_dice might be at epoch 50
- best_val_masd might be at epoch 70
- best_val_cldice might be at epoch 90

This creates a **natural source of model diversity** for ensembles — selecting ensemble members
by different metrics produces diverse checkpoints from the SAME training trajectory.

## Why This Matters for Ensembles

Mustafa et al. (2020) showed that **cross-entropy-based ensemble selection generalizes better
than accuracy-based selection** for small datasets:

> \"Ensembling to maximise cross-entropy: We initially constructed ensembles in order to maximise
> accuracy on the validation set. This resulted in multiple problems; with so few data points,
> ensembles were making decisions based on single data points (fighting for 0.5% improvement in
> accuracies on the 200 data point validation set). Furthermore, multiple ensembles generated by
> the greedy process would give the exact same validation accuracy, resulting in the need for an
> arbitrary tiebreaker.
>
> For all model sets, with upstream or downstream diversity, maximising cross entropy generalised
> better and avoided the latter problem, and was the most successful approach to avoiding
> over-fitting. As well as overall improving performance, we found it more reliable - when making
> ensembling decisions based on cross-entropy, final test performance monotonically increased with
> the inference budget (ensemble size). It is worth noting however that decisions made based on
> cross-entropy are worse for single models or smaller ensembles. With a small inference budget,
> we suggest optimising for accuracy. These phenomena are evident in Figure 13.
>
> Note that Caruana et al. (2004) suggest bagging of ensembles - i.e. selecting multiple ensembles
> using different splits of the data, then combining them in one pool. We did not opt for this, as
> we wanted to have a fixed downstream inference budget, and it is non-trivial to ensure that when
> combining multiple ensembles of variable size.\"

— Mustafa et al. (2020). "Deep Ensembles for Low-Data Transfer Learning." [arXiv:2010.06866](https://arxiv.org/abs/2010.06866)

![Ensembling metric selection from Mustafa et al. 2020](https://github.com/user-attachments/assets/placeholder)

## Current State

The multi-metric checkpoint tracking ALREADY WORKS in training:
- \`MultiMetricTracker\` in \`src/minivess/pipeline/multi_metric_tracker.py\`
- Saves 6-7 \`best_*.pth\` files per fold per run
- All uploaded to MLflow \`artifacts/checkpoints/\`

What's MISSING:
1. **Structured metadata** about which epoch each "best" came from
2. **Analysis flow integration** — ensemble construction should support
   "select members by metric X" as a strategy
3. **Factorial factor** — \`ensemble_metric_selection: [dice, cldice, masd, loss]\`

## Implementation

1. Add \`checkpoint_metadata.json\` to MLflow artifacts:
   \`\`\`json
   {
     "best_val_dice": {"epoch": 50, "value": 0.892},
     "best_val_cldice": {"epoch": 90, "value": 0.934},
     "best_val_masd": {"epoch": 70, "value": 2.1}
   }
   \`\`\`
2. Add ensemble strategy: \`per_metric_best\` — select ensemble members by
   different metrics (e.g., member 1 = best_dice, member 2 = best_cldice)
3. Compare ensemble-by-dice vs ensemble-by-cldice vs ensemble-by-loss

## References

- [Mustafa et al. (2020). "Deep Ensembles for Low-Data Transfer Learning." arXiv:2010.06866](https://arxiv.org/abs/2010.06866)
- Literature report: \`docs/planning/multi-metric-best-checkpoint-tracking-during-training-effect-for-ensembles.md\`
